### PR TITLE
Disconnect peer

### PR DESCRIPTION
**Pull request by afederigo**
Issues: #139

In this commit disconnect peer functionality have been added. By the given public key now we are able to disconnect the peer.

### CODE REVIEW

File: cmd/lncli/commands.go [hunk 1]
```diff
@@ -250,6 +250,40 @@ func connectPeer(ctx *cli.Context) error {
...
+}
+
+func disconnectPeer(ctx *cli.Context) error {
+	ctxb := context.Background()
+	client, cleanUp := getClient(ctx)
+	defer cleanUp()
+
+	targetAddress := ctx.Args().First()
+	splitAddr := strings.Split(targetAddress, "@")
+	if len(splitAddr) != 2 {
```
Comment (AndrewSamokhvalov) [inline cmd/lncli/commands.go]: With taking into account previous comment this logic should be simplified to parsing the `btcec.PublicKey`.

File: lnd_test.go [hunk 2]
```diff
@@ -236,6 +236,30 @@ func assertNumChannelsPending(ctxt context.Context, t *harnessTest,
 	}
 }
 
+//assertNumConnections asserts number current connections between two peers
+func assertNumConnections(ctxt context.Context, t *harnessTest,
+	alice, bob *lightningNode, expected int) {
+	aliceNumPeers, err := alice.ListPeers(ctxt, &lnrpc.ListPeersRequest{})
+	if err != nil {
+		t.Fatalf("error fetching alice's node (%v) list peers %v",
```
Comment (AndrewSamokhvalov) [inline lnd_test.go]: `error fetching` -> `unable to fetch`, the same logic should be applied to other error messages.

File: lnd_test.go [hunk 3]
```diff
@@ -236,6 +236,30 @@ func assertNumChannelsPending(ctxt context.Context, t *harnessTest,
...
+			alice.nodeID, err)
+	}
+	bobNumPeers, err := bob.ListPeers(ctxt, &lnrpc.ListPeersRequest{})
+	if err != nil {
+		t.Fatalf("error fetching bob's node (%v) list peers %v",
+			bob.nodeID, err)
+	}
+	if len(aliceNumPeers.Peers) != expected {
+		t.Fatalf("number of peers to alice connected incorrect. "+
```
Comment (AndrewSamokhvalov) [inline lnd_test.go]: Check the other tests to get right form of error explanation.

File: lnd_test.go [hunk 4]
```diff
@@ -291,6 +315,35 @@ func testBasicChannelFunding(net *networkHarness, t *harnessTest) {
...
 
+// testDisconnectingTargetPeer performs a test which
+// disconnects Alice-peer from Bob-peer and then re-connects them again
+func testDisconnectingTargetPeer(net *networkHarness, t *harnessTest) {
+
+	ctxb := context.Background()
+	timeout := time.Duration(time.Second * 10)
+	ctxt, _ := context.WithTimeout(ctxb, timeout)
+
+	// check existing connection.
```
Comment (AndrewSamokhvalov) [inline lnd_test.go]: To be consistent with the rest of the codebase, comments should start with uppercase, same should be applied to the rest of PR.

File: lnd_test.go [hunk 5]
```diff
@@ -291,6 +315,35 @@ func testBasicChannelFunding(net *networkHarness, t *harnessTest) {
...
+	if err != nil {
+		t.Fatalf("Disconnect peer: err %v", err)
+	}
+
+	// check zero peer connections.
+	assertNumConnections(ctxb, t, net.Alice, net.Bob, 0)
+
+	// Finally, re-connect both nodes.
+	if err := net.ConnectNodes(ctxt, net.Alice, net.Bob); err != nil {
+		t.Fatalf("344: Connect peer: err %v", err)
```
Comment (AndrewSamokhvalov) [inline lnd_test.go]: What `344` means?

File: lnrpc/README.md [hunk 6]
```diff
@@ -39,6 +39,8 @@ description):
        nested-pay-to-witness-key-hash (np2wkh).
   * ConnectPeer
      * Connects to a peer identified by a public key and host.
+  * DisconnectPeer
+     * Disconnects a remote peers identified by a public key and peerID.
```
Comment (AndrewSamokhvalov) [inline lnrpc/README.md]: > Disconnects a peer identified by a public key.

File: networktest.go [hunk 7]
```diff
@@ -662,6 +661,7 @@ func (n *networkHarness) SetUp() error {
 			return
 		}
 	}()
+
```
Comment (AndrewSamokhvalov) [inline networktest.go]: should be removed.

File: networktest.go [hunk 8]
```diff
@@ -974,7 +995,6 @@ func (n *networkHarness) OpenPendingChannel(ctx context.Context,
 		return nil, fmt.Errorf("unable to open channel between "+
 			"alice and bob: %v", err)
 	}
-
```
Comment (AndrewSamokhvalov) [inline networktest.go]: should be returned back.

File: server.go [hunk 9]
```diff
@@ -937,6 +947,56 @@ func (s *server) handleConnectPeer(msg *connectPeerMsg) {
 	}
 }
 
+// handleDisconnectPeer attempts to disconnect one peer from another
+func (s *server) handleDisconnectPeer(msg *disconnectPeerMsg) {
+	//targetPub := msg.pub_key
+	targetID := int32(msg.peerID)
+
+	//peers := s.Peers()
```
Comment (AndrewSamokhvalov) [inline server.go]: PR shouldn't contain the commented code.

File: server.go [hunk 10]
```diff
@@ -994,6 +1054,21 @@ func (s *server) ConnectToPeer(addr *lnwire.NetAddress,
 	return <-errChan
 }
 
+// DisconnectToPeer requests server for disconnect remote peer
```
Comment (AndrewSamokhvalov) [inline server.go]: > // DisconnectFromPeer sends the request to server to close the connection with peer 
// identified by public key.

File: rpcserver.go [hunk 11]
```diff
@@ -234,6 +234,27 @@ func (r *rpcServer) ConnectPeer(ctx context.Context,
 	return &lnrpc.ConnectPeerResponse{}, nil
 }
 
+// DisconnectPeer attempts to disconnect one peer from another
+// by a given pubKey or peerID.
+func (r *rpcServer) DisconnectPeer(ctx context.Context, in *lnrpc.DisconnectPeerRequest) (*lnrpc.DisconnectPeerResponse, error) {
+	//TODO(afederigo): remove switch
```
Comment (AndrewSamokhvalov) [inline rpcserver.go]: `TODO` should be accomplished, and removed.

File: lnd_test.go (hunk 5 above)
Comment (afederigo) [inline lnd_test.go]: It was a technical point (num of row in my IDLI), I forgot to remove.

**Comment by AndrewSamokhvalov** [comment]
Current implementation of `lnd` maintains the connection with peer if we have at least one opened channel. 

During local testing I have discovered that if we have the channel opened with peer and trying to disconnect with it, than the connection will be re-established because of functionality mentioned above. 

For that reason additional check that we does not have the established channels with remote peer should be added, and if we **have** than error message, that user should close established channels, should be returned.

**Comment by afederigo** [comment]
At the current time I have added some error explanation improvements, but checking open and pending channels before disconnect peer hasn't implemented yet.  
I have to get a trip about 10 days. 
After returning I'll be able to continue working on this issue.

**Comment by afederigo** [comment]
Some checks about existence pending or active channels while disconnecting peer has been added.

**Comment by AndrewSamokhvalov** [comment]
LGTM 👍

**Comment by afederigo** [comment]
I'm glad that my modest efforts were useful for the project 
I forgot about the persistent connection, (lost sight of it) sorry.
Your additional commits were very useful for me.
Thank you!

>Суббота,  6 мая 2017, 0:20 +03:00 от Olaoluwa Osuntokun <notifications@github.com>:
>
>@Roasbeef approved this pull request.
>LGTM 👾
>Some of the persistent connection logic shifted from under you throughout the lifetime of this PR. As a result I'll follow up the merging of this PR with a commit or two to ensure that if we formerly considered a peer as a persistent connection then disconnected them, we'd no longer attempt to reconnect automatically.
>However, on the receiving side of the disconnect if the peer still considers us a persistent connection, then they themselves will attempt to reconnect.
>—
>You are receiving this because you were mentioned.
>Reply to this email directly,  view it on GitHub , or  mute the thread .


Best regards,
Aleksandr

**Pull request merged**